src/compiler/scala/tools/nsc/typechecker/RefChecks.scala

```diff
@@ -686,15 +686,26 @@ abstract class RefChecks extends Transform {
               val matchingArity      = matchingName filter { m =>
                 !m.isDeferred &&
                 (m.name == underlying.name) &&
-                (m.paramLists.length == abstractParamLists.length) &&
-                (m.paramLists.map(_.length).sum == abstractParamLists.map(_.length).sum) &&
-                (m.tpe.typeParams.size == underlying.tpe.typeParams.size)
+                sameLength(m.paramLists, abstractParamLists) &&
+                sumSize(m.paramLists, 0) == sumSize(abstractParamLists, 0) &&
+                sameLength(m.tpe.typeParams, underlying.tpe.typeParams)
               }
 
               matchingArity match {
                 // So far so good: only one candidate method
                 case Scope(concrete)   =>
-                  val mismatches  = abstractParamLists.flatten.map(_.tpe) zip concrete.paramLists.flatten.map(_.tpe) filterNot { case (x, y) => x =:= y }
+                  //val mismatches  = abstractParamLists.flatten.map(_.tpe) zip concrete.paramLists.flatten.map(_.tpe) filterNot { case (x, y) => x =:= y }
+                  val mismatches: List[(Type, Type)] = {
+                    val res: ListBuffer[(Type, Type)] = ListBuffer.empty
+                    val aplIter = abstractParamLists .iterator.flatMap(x => x)
+                    val cplIter = concrete.paramLists.iterator.flatMap(x => x)
+                    while (res.length < 2 && aplIter.hasNext && cplIter.hasNext){
+                      val x = aplIter.next
+                      val y = cplIter.next
+                      if (!(x.tpe =:= y.tpe)) res += ((x.tpe, y.tpe))
+                    }
+                    res.toList
+                  }
```

Comment:
I'd prefer to stay with a functional style for readability.

How about:

```
val zipped: Iterator[(Symbol, Symbol)] = abstractParamLists .iterator.flatten.zip(concrete.paramLists.iterator.flatten)
zipped.filterNot { case (absParam, conParam) => absParam.tpe =:= conParam.tpe }.take(2).toList
```

Using `Iterator#{flatten,zip}` will avoid allocating temporary lists (although we will need to allocate tuples).

Reply:
Perhaps there is a better way: we can add to `Collections` a function to _zip_ two iterators: 
```Scala
zipIteratorsBy[A, B, C](itA: Iterator[A], itB: Iterator[B], fun: (A, B) => Option[C]): Iterator[C]
```
Which, in good functional style, would do just what you can guess from its signature.

Reply:
I'd be happy with the addition of:

```
/**  A more efficient version of `as.zip(bs).map( { case (a, b) => f(a, b) }`.
def zipWith[A, B, C](as: Iterator[A], bs: Iterator[B], fun: (A, B) => C): Iterator[C]
```

It's more efficient (no `Some` allocation), and easier to read IMO, to do the `filter` as a chained operation after calling this one.

Reply:
I already thought of using a `zipWith` iterator, but it would not be useful in this case. The specific problem of this case is that, if each pair passes the filter, then the pair should go through. 

Note that the `zipIteratorsBy` would only generate _at most_ two `Some` objects, and _at most_ two `Tuple2` objects.

Reply:
BTW, I already pushed a while ago the `mapFilter2` iterator I was thinking of.

Reply:
Makes sense.